src/sources/http_scrape/mod.rs

```diff
@@ -0,0 +1,233 @@
+        urls: Vec<Uri>,
+        interval_secs: u64,
+        headers: Option<HashMap<String, String>>,
+        auth: Option<Auth>,
+        tls: TlsSettings,
+        proxy: ProxyConfig,
+        shutdown: ShutdownSignal,
+    ) -> Self {
+        Self {
+            urls,
+            interval_secs,
+            headers,
+            auth,
+            tls,
+            proxy,
+            shutdown,
+        }
+    }
+}
+
+/// The default interval to scrape the http endpoint if none is configured.
+pub(crate) const fn default_scrape_interval_secs() -> u64 {
+    15
+}
+
+/// Methods that allow context-specific behavior during the scraping procedure.
+pub(crate) trait HttpScraper {
+    /// (Optional) Called before the HTTP request is made, allows building context.
+    fn build(&mut self, _url: &Uri) {}
+
+    /// Called after the HTTP request succeeds and returns the decoded/parsed Event array.
+    fn on_response(&mut self, url: &Uri, header: &Parts, body: &Bytes) -> Option<Vec<Event>>;
+
+    /// (Optional) Called if the HTTP response is not 200 ('OK').
+    fn on_http_response_error(&self, _uri: &Uri, _header: &Parts) {}
+}
+
+/// Builds a url for the HTTP requests.
+pub(crate) fn get_url(uri: &Uri, query: &Option<HashMap<String, Vec<String>>>) -> Uri {
+    let mut serializer = url::form_urlencoded::Serializer::new(String::new());
+    if let Some(query) = uri.query() {
+        serializer.extend_pairs(url::form_urlencoded::parse(query.as_bytes()));
+    };
+    if let Some(query) = &query {
+        for (k, l) in query {
+            for v in l {
+                serializer.append_pair(k, v);
+            }
+        }
+    };
+    let mut builder = Uri::builder();
+    if let Some(scheme) = uri.scheme() {
+        builder = builder.scheme(scheme.clone());
+    };
+    if let Some(authority) = uri.authority() {
+        builder = builder.authority(authority.clone());
+    };
+    builder = builder.path_and_query(match serializer.finish() {
+        query if !query.is_empty() => format!("{}?{}", uri.path(), query),
+        _ => uri.path().to_string(),
+    });
+    builder.build().expect("error building URI")
+}
+
+/// Scrapes one or more urls at an interval.
+///   - The HTTP request is built per the options in provided generic inputs.
+///   - The HTTP response is decoded/parsed into events by the specific context.
+///   - The events are then sent to the output stream.
+pub(crate) async fn http_scrape<H: HttpScraper + std::marker::Send + Clone>(
+    inputs: GenericHttpScrapeInputs,
+    context: H,
+    mut out: SourceSender,
+) -> Result<(), ()> {
+    let mut stream = IntervalStream::new(tokio::time::interval(Duration::from_secs(
+        inputs.interval_secs,
+    )))
+    .take_until(inputs.shutdown)
+    .map(move |_| stream::iter(inputs.urls.clone()))
+    .flatten()
+    .map(move |url| {
+        let client = HttpClient::new(inputs.tls.clone(), &inputs.proxy)
+            .expect("Building HTTP client failed");
```

Comment:
Under what conditions would this fail?

Reply:
Hmm good question. This came from the `prometheus_scrape` extracted code, and looking through the code base, I'm seeing basically everywhere it is used, its either an `unwrap()` or an `expect()` , but I'll take a stab at a comment for it. From a glance it looks like the failures would come from the application of the tls or proxy configuraitons.